File: packages/ash-go/deep_fuzzer_test.go
package ash

import (
	"fmt"
	"math/rand"
	"strings"
	"testing"
	"time"
	"unicode/utf8"
)

// ============================================================================
// DEEP FUZZER TESTS - High iteration fuzzing for robustness
// ============================================================================

func TestFuzzProofGeneration(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		nonce := randomHex(rng, 64)
		contextID := fmt.Sprintf("ctx_%d", i)
		binding := fmt.Sprintf("POST|/api/test/%d|", i)
		timestamp := fmt.Sprintf("%d", time.Now().Unix())
		bodyHash := HashBody(fmt.Sprintf(`{"i":%d}`, i))

		secret := DeriveClientSecret(nonce, contextID, binding)
		proof := BuildProofV21(secret, timestamp, binding, bodyHash)

		if len(proof) != 64 {
			t.Errorf("Iteration %d: proof length %d != 64", i, len(proof))
		}
		if !isHexString(proof) {
			t.Errorf("Iteration %d: proof not hex", i)
		}
	}
}

func TestFuzzProofVerification(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		nonce := randomHex(rng, 64)
		contextID := fmt.Sprintf("ctx_%d", i)
		binding := fmt.Sprintf("GET|/api/resource/%d|", i)
		timestamp := fmt.Sprintf("%d", time.Now().Unix())
		bodyHash := HashBody("")

		secret := DeriveClientSecret(nonce, contextID, binding)
		proof := BuildProofV21(secret, timestamp, binding, bodyHash)

		valid := VerifyProofV21(nonce, contextID, binding, timestamp, bodyHash, proof)
		if !valid {
			t.Errorf("Iteration %d: valid proof failed verification", i)
		}

		// Tamper with proof
		tamperedProof := proof[:len(proof)-1] + "0"
		if tamperedProof == proof {
			tamperedProof = proof[:len(proof)-1] + "1"
		}
		invalid := VerifyProofV21(nonce, contextID, binding, timestamp, bodyHash, tamperedProof)
		if invalid {
			t.Errorf("Iteration %d: tampered proof passed verification", i)
		}
	}
}

func TestFuzzJSONCanonicalizations(t *testing.T) {
	iterations := 1000

	for i := 0; i < iterations; i++ {
		// Test various JSON structures
		testCases := []interface{}{
			map[string]interface{}{"a": i, "b": i * 2},
			map[string]interface{}{"z": i, "a": i, "m": i},
			[]interface{}{i, i + 1, i + 2},
			map[string]interface{}{
				"nested": map[string]interface{}{
					"value": i,
				},
			},
		}

		for _, tc := range testCases {
			result1, err1 := CanonicalizeJSON(tc)
			result2, err2 := CanonicalizeJSON(tc)

			if err1 != nil || err2 != nil {
				continue // Skip invalid inputs
			}

			if result1 != result2 {
				t.Errorf("Iteration %d: non-deterministic canonicalization", i)
			}
		}
	}
}

func TestFuzzQueryCanonicalization(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		// Generate random query strings
		numParams := rng.Intn(10) + 1
		params := make([]string, numParams)
		for j := 0; j < numParams; j++ {
			key := fmt.Sprintf("key%d", rng.Intn(100))
			value := fmt.Sprintf("value%d", rng.Intn(100))
			params[j] = key + "=" + value
		}
		query := strings.Join(params, "&")

		result1, err1 := CanonicalizeQuery(query)
		result2, err2 := CanonicalizeQuery(query)

		if err1 != nil || err2 != nil {
			continue // Skip errors
		}

		if result1 != result2 {
			t.Errorf("Iteration %d: non-deterministic query canonicalization", i)
		}
	}
}

func TestFuzzBindingNormalization(t *testing.T) {
	methods := []string{"GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS", "HEAD"}
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		method := methods[rng.Intn(len(methods))]
		path := fmt.Sprintf("/api/v%d/resource/%d", rng.Intn(3)+1, i)
		query := fmt.Sprintf("page=%d&limit=%d", rng.Intn(100), rng.Intn(50))

		result1 := NormalizeBinding(method, path, query)
		result2 := NormalizeBinding(method, path, query)

		if result1 != result2 {
			t.Errorf("Iteration %d: non-deterministic binding normalization", i)
		}

		// Verify format
		parts := strings.Split(result1, "|")
		if len(parts) != 3 {
			t.Errorf("Iteration %d: binding should have 3 parts", i)
		}
	}
}

func TestFuzzHashBodyConsistency(t *testing.T) {
	iterations := 1000

	for i := 0; i < iterations; i++ {
		body := fmt.Sprintf(`{"index":%d,"data":"test_%d"}`, i, i*2)

		hash1 := HashBody(body)
		hash2 := HashBody(body)

		if hash1 != hash2 {
			t.Errorf("Iteration %d: non-deterministic hashing", i)
		}

		if len(hash1) != 64 {
			t.Errorf("Iteration %d: hash length %d != 64", i, len(hash1))
		}
	}
}

func TestFuzzClientSecretDerivation(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		nonce := randomHex(rng, 64)
		contextID := fmt.Sprintf("ash_ctx_%d", i)
		binding := fmt.Sprintf("POST|/api/action/%d|", i)

		secret1 := DeriveClientSecret(nonce, contextID, binding)
		secret2 := DeriveClientSecret(nonce, contextID, binding)

		if secret1 != secret2 {
			t.Errorf("Iteration %d: non-deterministic secret derivation", i)
		}

		if len(secret1) != 64 {
			t.Errorf("Iteration %d: secret length %d != 64", i, len(secret1))
		}
	}
}

func TestFuzzTimingSafeCompare(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		length := rng.Intn(100) + 10
		a := randomHex(rng, length)
		b := randomHex(rng, length)

		// Same strings should be equal
		if !TimingSafeCompare(a, a) {
			t.Errorf("Iteration %d: same string comparison failed", i)
		}

		// Different strings should not be equal (usually)
		if a != b && TimingSafeCompare(a, b) {
			t.Errorf("Iteration %d: different strings compared equal", i)
		}
	}
}

func TestFuzzScopedProofGeneration(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 500

	for i := 0; i < iterations; i++ {
		nonce := randomHex(rng, 64)
		contextID := fmt.Sprintf("ctx_%d", i)
		binding := "POST|/api/transfer|"
		timestamp := fmt.Sprintf("%d", time.Now().Unix())
		payload := map[string]interface{}{
			"amount":    i * 100,
			"recipient": fmt.Sprintf("user_%d", i),
			"memo":      fmt.Sprintf("test_%d", i),
		}
		scope := []string{"amount", "recipient"}

		secret := DeriveClientSecret(nonce, contextID, binding)
		result := BuildProofUnified(secret, timestamp, binding, payload, scope, "")

		if len(result.Proof) != 64 {
			t.Errorf("Iteration %d: proof length %d != 64", i, len(result.Proof))
		}

		if len(scope) > 0 && len(result.ScopeHash) != 64 {
			t.Errorf("Iteration %d: scope hash length %d != 64", i, len(result.ScopeHash))
		}
	}
}

func TestFuzzURLEncodedCanonicalization(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 1000

	for i := 0; i < iterations; i++ {
		numParams := rng.Intn(5) + 1
		params := make([]string, numParams)
		for j := 0; j < numParams; j++ {
			key := fmt.Sprintf("field%d", rng.Intn(20))
			value := fmt.Sprintf("val%d", rng.Intn(100))
			params[j] = key + "=" + value
		}
		input := strings.Join(params, "&")

		result1, err1 := CanonicalizeURLEncoded(input)
		result2, err2 := CanonicalizeURLEncoded(input)

		if err1 != nil || err2 != nil {
			continue // Skip errors
		}

		if result1 != result2 {
			t.Errorf("Iteration %d: non-deterministic URL encoding", i)
		}
	}
}

func TestFuzzUnicodeHandling(t *testing.T) {
	unicodeStrings := []string{
		"hello",
		"héllo",
		"日本語",
		"emoji 😀",
		"mixed café 日本",
		"special\t\n\r",
		"quotes\"and'stuff",
	}

	for i, s := range unicodeStrings {
		input := map[string]interface{}{"text": s}
		result1, err1 := CanonicalizeJSON(input)
		result2, err2 := CanonicalizeJSON(input)

		if err1 != nil || err2 != nil {
			continue
		}

		if result1 != result2 {
			t.Errorf("Unicode test %d: non-deterministic for %q", i, s)
		}

		if !utf8.ValidString(result1) {
			t.Errorf("Unicode test %d: invalid UTF-8 output", i)
		}
	}
}

func TestFuzzEmptyAndNullValues(t *testing.T) {
	testCases := []interface{}{
		map[string]interface{}{},
		map[string]interface{}{"empty": ""},
		map[string]interface{}{"null": nil},
		[]interface{}{},
		map[string]interface{}{"arr": []interface{}{}},
	}

	for i, tc := range testCases {
		result1, err1 := CanonicalizeJSON(tc)
		result2, err2 := CanonicalizeJSON(tc)

		if err1 != nil || err2 != nil {
			continue
		}

		if result1 != result2 {
			t.Errorf("Empty/null test %d: non-deterministic", i)
		}
	}
}

func TestFuzzLargePayloads(t *testing.T) {
	iterations := 100

	for i := 0; i < iterations; i++ {
		// Create large nested structure
		size := (i + 1) * 10
		data := make(map[string]interface{})
		for j := 0; j < size; j++ {
			data[fmt.Sprintf("key_%d", j)] = fmt.Sprintf("value_%d", j)
		}

		result1, err1 := CanonicalizeJSON(data)
		result2, err2 := CanonicalizeJSON(data)

		if err1 != nil || err2 != nil {
			continue
		}

		if result1 != result2 {
			t.Errorf("Large payload test %d: non-deterministic", i)
		}
	}
}

func TestFuzzSpecialCharactersInQuery(t *testing.T) {
	specialChars := []string{
		"a=1&b=2",
		"key=hello%20world",
		"name=John+Doe",
		"path=%2Ffoo%2Fbar",
		"unicode=%E6%97%A5%E6%9C%AC",
	}

	for i, query := range specialChars {
		result1, err1 := CanonicalizeQuery(query)
		result2, err2 := CanonicalizeQuery(query)

		if err1 != nil || err2 != nil {
			continue // Skip errors
		}

		if result1 != result2 {
			t.Errorf("Special char test %d: non-deterministic for %q", i, query)
		}
	}
}

func TestFuzzProofDifferentInputs(t *testing.T) {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	iterations := 500

	proofs := make(map[string]bool)

	for i := 0; i < iterations; i++ {
		nonce := randomHex(rng, 64)
		contextID := fmt.Sprintf("ctx_%d_%d", i, rng.Intn(1000))
		binding := fmt.Sprintf("POST|/api/%d|", i)
		timestamp := fmt.Sprintf("%d", time.Now().Unix()+int64(i))
		bodyHash := HashBody(fmt.Sprintf(`{"unique":%d}`, i))

		secret := DeriveClientSecret(nonce, contextID, binding)
		proof := BuildProofV21(secret, timestamp, binding, bodyHash)

		if proofs[proof] {
			t.Errorf("Iteration %d: duplicate proof generated", i)
		}
		proofs[proof] = true
	}
}

// Helper functions
func randomHex(rng *rand.Rand, length int) string {
	const hexChars = "0123456789abcdef"
	result := make([]byte, length)
	for i := range result {
		result[i] = hexChars[rng.Intn(16)]
	}
	return string(result)
}

func isHexString(s string) bool {
	for _, c := range s {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')) {
			return false
		}
	}
	return true
}

File: packages/ash-node/src/deep-fuzzer.test.ts
/**
 * ASH SDK Deep Fuzzer Test Suite
 *
 * High-iteration fuzzing with 10,000+ test cases per category.
 * Tests randomized inputs across all API surfaces.
 */

import { describe, it, expect, beforeAll } from 'vitest';
import * as fc from 'fast-check';
import * as crypto from 'crypto';
import {
  ashInit,
  ashBuildProofV21,
  ashVerifyProofV21,
  ashDeriveClientSecret,
  ashCanonicalizeJson,
  ashCanonicalizeUrlencoded,
  ashCanonicalizeQuery,
  ashNormalizeBinding,
  ashHashBody,
  ashGenerateNonce,
  ashGenerateContextId,
  canonicalizeJsonNative,
  canonicalQueryNative,
  ashBuildProofScoped,
  ashVerifyProofScoped,
  ashBuildProofUnified,
  ashVerifyProofUnified,
  ashExtractScopedFields,
  AshMemoryStore,
} from './index';

// High iteration counts for deep fuzzing
const FUZZ_RUNS = 10000;
const CRYPTO_RUNS = 5000;

beforeAll(() => {
  ashInit();
});

// Custom arbitraries
const hexString = (length: number) =>
  fc.array(fc.integer({ min: 0, max: 15 }), { minLength: length, maxLength: length })
    .map(arr => arr.map(n => n.toString(16)).join(''));

const alphaNumString = (minLen: number, maxLen: number) =>
  fc.stringMatching(new RegExp(`^[a-zA-Z0-9]{${minLen},${maxLen}}$`));

const alphaLowerString = (minLen: number, maxLen: number) =>
  fc.stringMatching(new RegExp(`^[a-z0-9]{${minLen},${maxLen}}$`));

const httpMethod = fc.constantFrom('GET', 'POST', 'PUT', 'DELETE', 'PATCH', 'HEAD', 'OPTIONS');

const urlPath = fc.array(alphaLowerString(1, 20), { minLength: 1, maxLength: 5 })
  .map(parts => '/' + parts.join('/'));

const queryParam = fc.tuple(
  alphaNumString(1, 20),
  alphaNumString(0, 50)
).map(([k, v]) => `${k}=${v}`);

const queryString = fc.array(queryParam, { minLength: 0, maxLength: 10 })
  .map(params => params.join('&'));

const jsonValue: fc.Arbitrary<unknown> = fc.letrec(tie => ({
  value: fc.oneof(
    fc.string(),
    fc.double({ noNaN: true, noDefaultInfinity: true }),
    fc.boolean(),
    fc.constant(null),
    fc.array(tie('value'), { maxLength: 5 }),
    fc.dictionary(fc.string({ minLength: 1, maxLength: 10 }), tie('value'), { maxKeys: 5 })
  )
})).value;

const jsonObject = fc.dictionary(
  fc.stringMatching(/^[a-zA-Z_]{1,20}$/),
  jsonValue,
  { minKeys: 1, maxKeys: 10 }
);

describe('DEEP FUZZER: Proof Generation/Verification (10,000 runs)', () => {
  it('property: verify(build(inputs)) === true', () => {
    fc.assert(
      fc.property(
        hexString(64),           // nonce (32 bytes = 64 hex)
        fc.stringMatching(/^[a-z0-9_]{4,40}$/).map(s => `ctx_${s}`),
        httpMethod,
        urlPath,
        queryString,
        hexString(64),           // bodyHash
        (nonce, contextId, method, path, query, bodyHash) => {
          const binding = ashNormalizeBinding(method, path, query);
          const timestamp = Math.floor(Date.now() / 1000).toString();

          const clientSecret = ashDeriveClientSecret(nonce, contextId, binding);
          const proof = ashBuildProofV21(clientSecret, timestamp, binding, bodyHash);

          // Property: what we build should verify
          const isValid = ashVerifyProofV21(nonce, contextId, binding, timestamp, bodyHash, proof);
          return isValid === true;
        }
      ),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: tampered proof never verifies', () => {
    fc.assert(
      fc.property(
        hexString(64),
        fc.stringMatching(/^[a-z0-9_]{4,20}$/).map(s => `ctx_${s}`),
        fc.integer({ min: 0, max: 63 }),  // tamper position
        (nonce, contextId, tamperPos) => {
          const binding = 'POST|/api/test|';
          const timestamp = Math.floor(Date.now() / 1000).toString();
          const bodyHash = crypto.randomBytes(32).toString('hex');

          const clientSecret = ashDeriveClientSecret(nonce, contextId, binding);
          const proof = ashBuildProofV21(clientSecret, timestamp, binding, bodyHash);

          // Tamper at random position
          const chars = proof.split('');
          chars[tamperPos] = chars[tamperPos] === 'a' ? 'b' : 'a';
          const tamperedProof = chars.join('');

          if (tamperedProof === proof) return true; // Skip if no change

          const isValid = ashVerifyProofV21(nonce, contextId, binding, timestamp, bodyHash, tamperedProof);
          return isValid === false;
        }
      ),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: different nonce produces different proof', () => {
    fc.assert(
      fc.property(
        hexString(64),
        hexString(64),
        (nonce1, nonce2) => {
          if (nonce1 === nonce2) return true; // Skip identical

          const contextId = 'ctx_test';
          const binding = 'POST|/api/test|';
          const timestamp = Math.floor(Date.now() / 1000).toString();
          const bodyHash = crypto.randomBytes(32).toString('hex');

          const secret1 = ashDeriveClientSecret(nonce1, contextId, binding);
          const secret2 = ashDeriveClientSecret(nonce2, contextId, binding);

          const proof1 = ashBuildProofV21(secret1, timestamp, binding, bodyHash);
          const proof2 = ashBuildProofV21(secret2, timestamp, binding, bodyHash);

          return proof1 !== proof2;
        }
      ),
      { numRuns: CRYPTO_RUNS }
    );
  });
});

describe('DEEP FUZZER: JSON Canonicalization (10,000 runs)', () => {
  it('property: canonicalization is deterministic', () => {
    fc.assert(
      fc.property(jsonObject, (obj) => {
        const json = JSON.stringify(obj);
        try {
          const canon1 = canonicalizeJsonNative(json);
          const canon2 = canonicalizeJsonNative(json);
          return canon1 === canon2;
        } catch {
          return true; // Skip invalid JSON
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: canonicalization is idempotent', () => {
    fc.assert(
      fc.property(jsonObject, (obj) => {
        const json = JSON.stringify(obj);
        try {
          const canon1 = canonicalizeJsonNative(json);
          const canon2 = canonicalizeJsonNative(canon1);
          return canon1 === canon2;
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: key order does not affect canonical form', () => {
    fc.assert(
      fc.property(
        fc.dictionary(fc.string({ minLength: 1, maxLength: 10 }), fc.string(), { minKeys: 2, maxKeys: 5 }),
        (obj) => {
          const keys = Object.keys(obj);
          if (keys.length < 2) return true;

          // Create two objects with different key orders
          const obj1: Record<string, string> = {};
          const obj2: Record<string, string> = {};

          keys.forEach(k => obj1[k] = obj[k]);
          keys.reverse().forEach(k => obj2[k] = obj[k]);

          try {
            const canon1 = canonicalizeJsonNative(JSON.stringify(obj1));
            const canon2 = canonicalizeJsonNative(JSON.stringify(obj2));
            return canon1 === canon2;
          } catch {
            return true;
          }
        }
      ),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: canonical JSON is valid JSON', () => {
    fc.assert(
      fc.property(jsonObject, (obj) => {
        const json = JSON.stringify(obj);
        try {
          const canonical = canonicalizeJsonNative(json);
          JSON.parse(canonical);
          return true;
        } catch {
          return true; // Skip errors
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });
});

describe('DEEP FUZZER: Query String Canonicalization (10,000 runs)', () => {
  it('property: canonicalization is deterministic', () => {
    fc.assert(
      fc.property(queryString, (query) => {
        try {
          const canon1 = canonicalQueryNative(query);
          const canon2 = canonicalQueryNative(query);
          return canon1 === canon2;
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: canonicalization is idempotent', () => {
    fc.assert(
      fc.property(queryString, (query) => {
        try {
          const canon1 = canonicalQueryNative(query);
          const canon2 = canonicalQueryNative(canon1);
          return canon1 === canon2;
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: parameter order does not affect canonical form', () => {
    fc.assert(
      fc.property(
        fc.array(queryParam, { minLength: 2, maxLength: 5 }),
        (params) => {
          const query1 = params.join('&');
          const query2 = [...params].reverse().join('&');

          try {
            const canon1 = canonicalQueryNative(query1);
            const canon2 = canonicalQueryNative(query2);
            return canon1 === canon2;
          } catch {
            return true;
          }
        }
      ),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: result is sorted by key', () => {
    fc.assert(
      fc.property(queryString, (query) => {
        try {
          const canonical = canonicalQueryNative(query);
          if (!canonical) return true;

          const pairs = canonical.split('&');
          for (let i = 1; i < pairs.length; i++) {
            const prevKey = pairs[i-1].split('=')[0];
            const currKey = pairs[i].split('=')[0];
            if (Buffer.from(prevKey).compare(Buffer.from(currKey)) > 0) {
              return false;
            }
          }
          return true;
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });
});

describe('DEEP FUZZER: Binding Normalization (10,000 runs)', () => {
  it('property: normalization is deterministic', () => {
    fc.assert(
      fc.property(httpMethod, urlPath, queryString, (method, path, query) => {
        try {
          const norm1 = ashNormalizeBinding(method, path, query);
          const norm2 = ashNormalizeBinding(method, path, query);
          return norm1 === norm2;
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: method is always uppercase', () => {
    fc.assert(
      fc.property(httpMethod, urlPath, (method, path) => {
        try {
          const normalized = ashNormalizeBinding(method.toLowerCase(), path, '');
          return normalized.startsWith(method.toUpperCase() + '|');
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });

  it('property: format is METHOD|PATH|QUERY', () => {
    fc.assert(
      fc.property(httpMethod, urlPath, queryString, (method, path, query) => {
        try {
          const normalized = ashNormalizeBinding(method, path, query);
          const parts = normalized.split('|');
          return parts.length === 3;
        } catch {
          return true;
        }
      }),
      { numRuns: FUZZ_RUNS }
    );
  });
});

describe('DEEP FUZZER: Cryptographic Properties (5,000 runs)', () => {
  it('property: client secrets are unique for different inputs', () => {
    const secrets = new Set<string>();

    fc.assert(
      fc.property(
        hexString(64),
        fc.stringMatching(/^[a-z0-9]{5,20}$/).map(s => `ctx_${s}`),
        (nonce, contextId) => {
          const binding = 'POST|/api/test|';
          const secret = ashDeriveClientSecret(nonce, contextId, binding);

          if (secrets.has(secret)) {
            return false; // Collision detected!
          }
          secrets.add(secret);
          return true;
        }
      ),
      { numRuns: CRYPTO_RUNS }
    );
  });

  it('property: body hashes are unique for different inputs', () => {
    const hashes = new Set<string>();

    fc.assert(
      fc.property(fc.string({ minLength: 1, maxLength: 1000 }), (body) => {
        const hash = ashHashBody(body);

        // Note: We don't check for collisions here as different strings
        // could theoretically produce the same hash (though extremely unlikely)
        // Instead we verify the hash is always 64 hex chars
        return hash.length === 64 && /^[0-9a-f]+$/.test(hash);
      }),
      { numRuns: CRYPTO_RUNS }
    );
  });

  it('property: proofs are exactly 64 hex characters', () => {
    fc.assert(
      fc.property(hexString(64), (nonce) => {
        const contextId = 'ctx_test';
        const binding = 'POST|/api/test|';
        const timestamp = Math.floor(Date.now() / 1000).toString();
        const bodyHash = crypto.randomBytes(32).toString('hex');

        const clientSecret = ashDeriveClientSecret(nonce, contextId, binding);
        const proof = ashBuildProofV21(clientSecret, timestamp, binding, bodyHash);

        return proof.length === 64 && /^[0-9a-f]+$/.test(proof);
      }),
      { numRuns: CRYPTO_RUNS }
    );
  });

  it('property: nonces are cryptographically random', () => {
    const nonces: string[] = [];

    // Generate many nonces
    for (let i = 0; i < 1000; i++) {
      nonces.push(ashGenerateNonce());
    }

    // Check uniqueness
    const unique = new Set(nonces);
    expect(unique.size).toBe(nonces.length);

    // Check entropy distribution (chi-squared test approximation)
    const charCounts = new Map<string, number>();
    for (const nonce of nonces) {
      for (const char of nonce) {
        charCounts.set(char, (charCounts.get(char) || 0) + 1);
      }
    }

    // All 16 hex chars should appear
    expect(charCounts.size).toBe(16);

    // Distribution should be roughly uniform
    const counts = Array.from(charCounts.values());
    const mean = counts.reduce((a, b) => a + b, 0) / counts.length;
    const variance = counts.reduce((sum, c) => sum + Math.pow(c - mean, 2), 0) / counts.length;
    const coefficientOfVariation = Math.sqrt(variance) / mean;

    // CV should be low for uniform distribution
    expect(coefficientOfVariation).toBeLessThan(0.15);
  });
});

// Dangerous keys that are blocked by the SDK
const DANGEROUS_KEYS = ['__proto__', 'constructor', 'prototype'];

describe('DEEP FUZZER: Scoped Proofs (5,000 runs)', () => {
  it('property: scoped proof verifies with correct scope', () => {
    fc.assert(
      fc.property(
        hexString(64),
        fc.stringMatching(/^[a-z0-9]{4,20}$/).map(s => `ctx_${s}`),
        fc.dictionary(
          fc.stringMatching(/^[a-z]{1,10}$/),
          fc.oneof(fc.string(), fc.integer(), fc.boolean()),
          { minKeys: 2, maxKeys: 5 }
        ),
        (nonce, contextId, payload) => {
          // Filter out dangerous keys that would be rejected
          const safeKeys = Object.keys(payload).filter(k => !DANGEROUS_KEYS.includes(k));
          if (safeKeys.length < 1) return true;

          // Build safe payload
          const safePayload: Record<string, unknown> = {};
          for (const k of safeKeys) {
            safePayload[k] = payload[k];
          }

          const scope = safeKeys.slice(0, Math.ceil(safeKeys.length / 2));
          const binding = 'POST|/api/scoped|';
          const timestamp = Math.floor(Date.now() / 1000).toString();

          const clientSecret = ashDeriveClientSecret(nonce, contextId, binding);
          const { proof, scopeHash } = ashBuildProofScoped(
            clientSecret, timestamp, binding, safePayload, scope
          );

          const isValid = ashVerifyProofScoped(
            nonce, contextId, binding, timestamp, safePayload, scope, scopeHash, proof
          );

          return isValid === true;
        }
      ),
      { numRuns: CRYPTO_RUNS }
    );
  });

  it('property: scoped proof fails with wrong scope', () => {
    fc.assert(
      fc.property(
        hexString(64),
        fc.stringMatching(/^[a-z0-9]{4,20}$/).map(s => `ctx_${s}`),
        (nonce, contextId) => {
          const payload = { amount: 100, recipient: 'alice', memo: 'test' };
          const scope = ['amount', 'recipient'];
          const wrongScope = ['amount', 'memo'];

          const binding = 'POST|/api/transfer|';
          const timestamp = Math.floor(Date.now() / 1000).toString();

          const clientSecret = ashDeriveClientSecret(nonce, contextId, binding);
          const { proof, scopeHash } = ashBuildProofScoped(
            clientSecret, timestamp, binding, payload, scope
          );

          const isValid = ashVerifyProofScoped(
            nonce, contextId, binding, timestamp, payload, wrongScope, scopeHash, proof
          );

          return isValid === false;
        }
      ),
      { numRuns: CRYPTO_RUNS }
    );
  });
});

describe('DEEP FUZZER: Context Store Race Conditions (1,000 runs)', () => {
  it('property: exactly one concurrent consume succeeds', async () => {
    const results: boolean[] = [];

    for (let i = 0; i < 100; i++) {
      const store = new AshMemoryStore();
      const ctx = await store.create({
        binding: 'POST|/api/test|',
        ttlMs: 60000,
      });

      // Launch 10 concurrent consumes
      const consumeResults = await Promise.all(
        Array(10).fill(null).map(() => store.consume(ctx.id))
      );

      const successCount = consumeResults.filter(r => r === true).length;
      results.push(successCount === 1);
    }

    // All iterations should have exactly 1 success
    expect(results.every(r => r)).toBe(true);
  });
});

describe('DEEP FUZZER: Error Handling (5,000 runs)', () => {
  it('property: invalid nonce format throws', () => {
    fc.assert(
      fc.property(
        fc.string({ minLength: 1, maxLength: 100 }),
        (invalidNonce) => {
          // Skip valid hex strings of correct length
          if (/^[0-9a-fA-F]{64,128}$/.test(invalidNonce)) return true;

          try {
            ashDeriveClientSecret(invalidNonce, 'ctx_test', 'GET|/|');
            return false; // Should have thrown
          } catch {
            return true;
          }
        }
      ),
      { numRuns: CRYPTO_RUNS }
    );
  });

  it('property: empty inputs throw', () => {
    const emptyTests = [
      () => ashDeriveClientSecret('', 'ctx', 'GET|/|'),
      () => ashDeriveClientSecret(crypto.randomBytes(32).toString('hex'), '', 'GET|/|'),
      () => ashDeriveClientSecret(crypto.randomBytes(32).toString('hex'), 'ctx', ''),
    ];

    for (const test of emptyTests) {
      expect(() => test()).toThrow();
    }
  });

  it('property: oversized inputs are rejected', () => {
    // Large JSON should be rejected
    const largeJson = JSON.stringify({ data: 'x'.repeat(11 * 1024 * 1024) });
    expect(() => canonicalizeJsonNative(largeJson)).toThrow();

    // Oversized nonce should be rejected
    const longNonce = crypto.randomBytes(100).toString('hex');
    expect(() => ashDeriveClientSecret(longNonce, 'ctx', 'GET|/|')).toThrow();
  });
});

console.log('Deep Fuzzer Test Suite loaded - 10,000+ iterations per test');
